File: java/iforevents/src/main/java/com/iforevents/Iforevents.java
package com.iforevents;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Callable;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * The facade: one {@link #init()}, then {@link #identify}, {@link #track},
 * {@link #page}, {@link #reset}, {@link #flush} and {@link #shutdown} fan out
 * to every integration in isolation. Identify traits are remembered and
 * merged under later track properties; nested maps are flattened with
 * {@code _}. Mirrors the {@code Iforevents} class of the Flutter package.
 */
public class Iforevents {
    /** Supplies platform context merged into identify traits. */
    public interface ContextProvider {
        Map<String, Object> get();
    }

    /** Observes the outcome of every fan-out. */
    public interface ResultObserver {
        void onResults(List<IntegrationResult> results);
    }

    private static final Logger LOG = Logger.getLogger("iforevents");

    private final List<Integration> integrations = new ArrayList<Integration>();
    private final ContextProvider context;
    private final boolean debug;
    private final ResultObserver observer;
    private volatile Map<String, Object> traits = Collections.emptyMap();
    private volatile boolean initialized;

    private Iforevents(Builder b) {
        this.integrations.addAll(b.integrations);
        this.context = b.context;
        this.debug = b.debug;
        this.observer = b.observer;
    }

    public static Builder builder() {
        return new Builder();
    }

    public static final class Builder {
        private final List<Integration> integrations = new ArrayList<Integration>();
        private ContextProvider context = new ContextProvider() {
            @Override
            public Map<String, Object> get() {
                return Context.defaultContext();
            }
        };
        private boolean debug;
        private ResultObserver observer;

        public Builder integration(Integration i) { integrations.add(i); return this; }
        public Builder integrations(List<? extends Integration> list) { integrations.addAll(list); return this; }
        public Builder context(ContextProvider c) { context = c; return this; }
        public Builder debug(boolean d) { debug = d; return this; }
        public Builder resultObserver(ResultObserver o) { observer = o; return this; }
        public Iforevents build() { return new Iforevents(this); }
    }

    public boolean isInitialized() { return initialized; }

    /** A copy of the traits remembered from the last identify. */
    public Map<String, Object> currentTraits() { return new LinkedHashMap<String, Object>(traits); }

    public synchronized void addIntegration(Integration i) { integrations.add(i); }

    public Integration integration(String name) {
        for (Integration i : integrations) if (i.name().equals(name)) return i;
        return null;
    }

    /** Initializes every integration. Failures are reported, never thrown. */
    public List<IntegrationResult> init() {
        List<IntegrationResult> results = fanOut(new Action() {
            @Override
            public void run(Integration i) throws Exception { i.init(); }
        });
        initialized = true;
        return results;
    }

    public List<IntegrationResult> identify(final String customId, Map<String, ?> traits) {
        if (customId == null || customId.isEmpty() || !ready("identify")) return Collections.emptyList();
        Map<String, Object> merged = new LinkedHashMap<String, Object>(safeContext());
        if (traits != null) merged.putAll(traits);
        final Map<String, Object> flat = Flatten.flatten(merged);
        List<IntegrationResult> results = fanOut(new Action() {
            @Override
            public void run(Integration i) throws Exception { i.identify(new IdentifyEvent(customId, flat)); }
        });
        this.traits = flat;
        return results;
    }

    public List<IntegrationResult> track(String name, Map<String, ?> properties) {
        if (name == null || name.isEmpty() || !ready("track")) return Collections.emptyList();
        Map<String, Object> merged = new LinkedHashMap<String, Object>(traits);
        if (properties != null) merged.putAll(properties);
        final TrackEvent event = new TrackEvent(name, Flatten.flatten(merged));
        return fanOut(new Action() {
            @Override
            public void run(Integration i) throws Exception { i.track(event); }
        });
    }

    public List<IntegrationResult> track(String name) { return track(name, null); }

    public List<IntegrationResult> page(String name, Map<String, ?> properties) { return page(name, properties, null, null, null); }

    public List<IntegrationResult> page(String name, Map<String, ?> properties, String navigationType, String toRoute, String previousRoute) {
        if (!ready("page")) return Collections.emptyList();
        final PageEvent event = new PageEvent(name, Flatten.flatten(properties), navigationType, toRoute, previousRoute);
        return fanOut(new Action() {
            @Override
            public void run(Integration i) throws Exception { i.page(event); }
        });
    }

    /** {@link #page} with mobile naming. */
    public List<IntegrationResult> screen(String name, Map<String, ?> properties) { return page(name, properties); }

    /** Forgets the user in every integration (logout). */
    public List<IntegrationResult> reset() {
        if (!ready("reset")) return Collections.emptyList();
        List<IntegrationResult> results = fanOut(new Action() {
            @Override
            public void run(Integration i) throws Exception { i.reset(); }
        });
        traits = Collections.emptyMap();
        return results;
    }

    public List<IntegrationResult> flush() {
        return fanOut(new Action() {
            @Override
            public void run(Integration i) throws Exception { i.flush(); }
        });
    }

    /** Flushes and releases every integration; the instance is no longer usable. */
    public List<IntegrationResult> shutdown() {
        List<IntegrationResult> results = fanOut(new Action() {
            @Override
            public void run(Integration i) throws Exception { i.shutdown(); }
        });
        initialized = false;
        return results;
    }

    private interface Action {
        void run(Integration i) throws Exception;
    }

    private boolean ready(String method) {
        if (initialized) return true;
        if (debug) LOG.log(Level.WARNING, "[iforevents] " + method + " called before init; ignored");
        return false;
    }

    private Map<String, Object> safeContext() {
        try {
            Map<String, Object> c = context == null ? null : context.get();
            return c == null ? Collections.<String, Object>emptyMap() : c;
        } catch (RuntimeException e) {
            if (debug) LOG.log(Level.WARNING, "[iforevents] context provider failed", e);
            return Collections.emptyMap();
        }
    }

    /** Runs one integration call in isolation and reports the outcome. */
    public static IntegrationResult safeExecute(Integration integration, Callable<Void> action) {
        try {
            action.call();
            return new IntegrationResult(integration.name(), true, null);
        } catch (Throwable t) {
            return new IntegrationResult(integration.name(), false, t);
        }
    }

    private List<IntegrationResult> fanOut(final Action action) {
        List<Integration> snapshot;
        synchronized (this) {
            snapshot = new ArrayList<Integration>(integrations);
        }
        List<IntegrationResult> results = new ArrayList<IntegrationResult>(snapshot.size());
        for (final Integration i : snapshot) {
            IntegrationResult r = safeExecute(i, new Callable<Void>() {
                @Override
                public Void call() throws Exception {
                    action.run(i);
                    return null;
                }
            });
            if (!r.success && debug) LOG.log(Level.WARNING, "[iforevents] " + r.integration + " failed", r.error);
            results.add(r);
        }
        if (observer != null) observer.onResults(results);
        return results;
    }
}

File: js/packages/core/src/iforevents.ts
import { defaultContext } from "./context";
import { flatten } from "./flatten";
import { Integration, safeExecute } from "./integration";
import type { ContextProvider, IntegrationResult, Logger, PageEvent, Properties } from "./types";

export interface IforeventsOptions {
  /** Integrations to fan every call out to, in order. */
  integrations?: Integration[];
  /** Platform context merged into identify traits (device, sdk, runtime). */
  context?: ContextProvider;
  /** Log integration failures. Default false. */
  debug?: boolean;
  logger?: Logger;
  /** Observe the outcome of every fan-out (for retries, metrics, tests). */
  onResult?: (results: IntegrationResult[]) => void;
}

export interface PageOptions {
  navigationType?: string;
  toRoute?: string;
  previousRoute?: string;
}

/**
 * The facade: one `init`, then `identify`, `track`, `page`, `reset`, `flush`
 * and `shutdown` fan out to every integration in isolation. Identify traits
 * are remembered and merged under later track properties; nested objects are
 * flattened with `_`. Mirrors the `Iforevents` class of the Flutter package.
 */
export class Iforevents {
  private integrations: Integration[] = [];
  private traits: Properties = {};
  private initialized = false;
  private contextProvider: ContextProvider;
  private readonly logger: Logger | null;
  private readonly onResult?: (results: IntegrationResult[]) => void;

  constructor(options: IforeventsOptions = {}) {
    this.integrations = [...(options.integrations ?? [])];
    this.contextProvider = options.context ?? defaultContext;
    this.logger = options.debug ? (options.logger ?? console) : null;
    this.onResult = options.onResult;
  }

  get isInitialized(): boolean {
    return this.initialized;
  }

  /** Traits remembered from the last identify (context included). */
  get currentTraits(): Properties {
    return { ...this.traits };
  }

  /** Registers an integration; call `init` (or `initIntegration`) afterwards. */
  addIntegration(integration: Integration): this {
    this.integrations.push(integration);
    return this;
  }

  /** Finds a registered integration by name or class. */
  getIntegration<T extends Integration>(match: string | (new (...args: never[]) => T)): T | undefined {
    return this.integrations.find((i) => (typeof match === "string" ? i.name === match : i instanceof match)) as T | undefined;
  }

  /** Initializes every integration. Failures are reported, never thrown. */
  async init(options: Pick<IforeventsOptions, "integrations" | "context"> = {}): Promise<IntegrationResult[]> {
    if (options.integrations) this.integrations.push(...options.integrations);
    if (options.context) this.contextProvider = options.context;
    const results = await this.fanOut((i) => i.init());
    this.initialized = true;
    return results;
  }

  async identify(customId: string, traits: Properties = {}): Promise<IntegrationResult[]> {
    if (!customId) return [];
    if (!this.ready("identify")) return [];
    const context = await this.safeContext();
    const merged = flatten({ ...context, ...traits });
    const results = await this.fanOut((i) => i.identify({ customId, traits: merged }));
    this.traits = merged;
    return results;
  }

  async track(name: string, properties: Properties = {}): Promise<IntegrationResult[]> {
    if (!name) return [];
    if (!this.ready("track")) return [];
    const merged = flatten({ ...this.traits, ...properties });
    const timestamp = new Date();
    return this.fanOut((i) => i.track({ name, type: "track", properties: merged, timestamp }));
  }

  async page(name?: string, properties: Properties = {}, options: PageOptions = {}): Promise<IntegrationResult[]> {
    if (!this.ready("page")) return [];
    const event: PageEvent = { name: name || "page_view", properties: flatten(properties), timestamp: new Date(), ...options };
    return this.fanOut((i) => i.page(event));
  }

  /** Alias of `page` for mobile-style naming. */
  screen(name?: string, properties: Properties = {}, options: PageOptions = {}): Promise<IntegrationResult[]> {
    return this.page(name, properties, options);
  }

  /** Forgets the user in every integration (logout). */
  async reset(): Promise<IntegrationResult[]> {
    if (!this.ready("reset")) return [];
    const results = await this.fanOut((i) => i.reset());
    this.traits = {};
    return results;
  }

  async flush(): Promise<IntegrationResult[]> {
    return this.fanOut((i) => i.flush());
  }

  /** Flushes and releases every integration; the instance is no longer usable. */
  async shutdown(): Promise<IntegrationResult[]> {
    const results = await this.fanOut((i) => i.shutdown());
    this.initialized = false;
    return results;
  }

  private ready(method: string): boolean {
    if (this.initialized) return true;
    this.logger?.warn(`[iforevents] ${method} called before init; ignored`);
    return false;
  }

  private async safeContext(): Promise<Properties> {
    try {
      return (await this.contextProvider()) ?? {};
    } catch (error) {
      this.logger?.warn("[iforevents] context provider failed", error);
      return {};
    }
  }

  private async fanOut(action: (integration: Integration) => Promise<void> | void): Promise<IntegrationResult[]> {
    const results = await Promise.all(this.integrations.map((integration) => safeExecute(integration, () => action(integration))));
    for (const result of results) {
      if (!result.success) this.logger?.warn(`[iforevents] ${result.integration} failed`, result.error);
    }
    this.onResult?.(results);
    return results;
  }
}
